api/types/access_list.go

```diff
@@ -0,0 +1,274 @@
+	// For a membership to an access list to be effective, the user must meet the requirements of
+	// MembershipRequires and must be in the members list.
+	MembershipRequires *AccessListRequires `json:"membership_requires" yaml:"membership_requires"`
+
+	// OwnershipRequires describes the requirements for a user to be an owner of the access list.
+	// For ownership of an access list to be effective, the user must meet the requirements of
+	// OwnershipRequires and must be in the owners list.
+	OwnershipRequires *AccessListRequires `json:"ownership_requires" yaml:"ownership_requires"`
+
+	// Grants describes the access granted by membership to this access list.
+	Grants *AccessListGrants `json:"grants" yaml:"grants"`
+
+	// Members describes the current members of the access list.
+	Members []*AccessListMember `json:"members" yaml:"members"`
+}
+
+// AccessListOwner is an owner of an access list.
+type AccessListOwner struct {
+	// Name is the username of the owner.
+	Name string `json:"name" yaml:"name"`
+
+	// Description is the plaintext description of the owner and why they are an owner.
+	Description string `json:"description" yaml:"description"`
+}
+
+// AccessListAudit describes the audit configuration for an access list.
+type AccessListAudit struct {
+	// Frequency is a duration that describes how often an access list must be audited.
+	Frequency time.Duration `json:"frequency" yaml:"frequency"`
+}
+
+// AccessListRequires describes a requirement section for an access list. A user must
+// meet the following criteria to obtain the specific access to the list.
+type AccessListRequires struct {
+	// Roles are the user roles that must be present for the user to obtain access.
+	Roles []string `json:"roles" yaml:"roles"`
+
+	// Traits are the traits that must be present for the user to obtain access.
+	Traits map[string][]string `json:"traits" yaml:"traits"`
+}
+
+// AccessListGrants describes what access is granted by membership to the access list.
+type AccessListGrants struct {
+	// Roles are the roles that are granted to users who are members of the access list.
+	Roles []string `json:"roles" yaml:"roles"`
+
+	// Traits are the traits that are granted to users who are members of the access list.
+	Traits map[string][]string `json:"traits" yaml:"traits"`
+}
+
+// AccessListMember describes a member of an access list.
+type AccessListMember struct {
+	// Name is the name of the member of the access list.
+	Name string `json:"name" yaml:"name"`
+
+	// Joined is when the user joined the access list.
+	Joined time.Time `json:"joined" yaml:"joined"`
+
+	// expires is when the user's membership to the access list expires.
+	Expires time.Time `json:"expires" yaml:"expires"`
+
+	// reason is the reason this user was added to the access list.
+	Reason string `json:"reason" yaml:"reason"`
+
+	// added_by is the user that added this user to the access list.
+	AddedBy string `json:"added_by" yaml:"added_by"`
```

Comment:
I _think_ the YAML package should leverage JSON tags if no YAML tag is present. Might be worth double checking.

Reply:
It doesn't look like it does. We use this: https://github.com/ghodss/yaml under the covers in order to do this. That means that these `yaml` tags won't necessarily do anything immediately, but IMO I'd love to get rid of `ghodss` given that it's most recent release is 6 years old. So I'd love to keep these included for now.